anndata/tests/test_layers.py

```diff
@@ -41,21 +41,30 @@ def test_views():
     assert not adata_view.isview
 
 
+def test_set_dataframe():
+    adata = AnnData(X)
+    df = gen_typed_df_t2_size(*X.shape)
+    with pytest.warns(UserWarning, match=r"Layer 'df'.*dtype object"):
+        adata.layers["df"] = df
+    assert isinstance(adata.layers["df"], np.ndarray)
+    assert np.issubdtype(adata.layers["df"].dtype, object)
```

Comment:
Please add cases for: `obsp`, `varp`, and a homogenous dataframe.

Also, does this do the same thing as `adata.X = df`?

Reply:
will do